### ISSUE TITLE
Ideas for iNDS 2.0.0

### ISSUE THREAD

**Issue by DarknesGaming** [WilliamLCobb/iNDS#121]
# Ideas for iNDS 2.0.0
--------------------------------------------------------------------------------
Here is the list for what should be added in iNDS 2.0.0, followed by the fact that my sandwich needs more mayonnaise. I practically searched the ENTIRE list of open issues finding them.
- Anything from issue #85 by Anuskuss.
- Remaking the iNDS interface by @WilliamLCobb 
- 3D rendering (high-resolution sounds good)
- ARM JIT support for Jailed devices (not complete yet)
- If possible, more filters.
- A tvOS version.
- Multiple displays on each screen, which is a reference to #92 
- Stop the audio lag
- Being able to rename ROMs (#69)
- That's all I got.

**Comment by shiun1234** [comment]
Hope the new version iPhone 5 can be full speed👍🏻

**Comment by OkazakiTheOtaku** [comment]
I don't think the iPhone 5 can ever be full speed. It's simply not powerful enough. It's not even 64 bit.

**Comment by inpuu** [comment]
"ARM JIT support for Jailed devices" 
Is this possible?! If it is really possible, it will be GREAT!!!!!

**Comment by pmp174** [comment]
A new interface thats easier to navigate also

**Comment by hector0406** [comment]
1.5x speed 💃

**Comment by DarknesGaming** [comment]
@inpuu ARM JIT is possible. It will work for both 32-bit and 64-bit devices soon, but that will take some time. It might be faster than the DeSmuME 0.9.10 Threaded ARM Interpreter.
@shiun1234 the iPhone 5 is a 32-bit device, so it can't run iNDS at full speed. And before you ask, the iPhone 5C uses the same CPU and is also 32-bit, so that won't run at full speed, either. Any device using the Apple A6/A6X CPU or below do not have the 64-bit architecture needed to run at full speed. nds4ios can prove that without even trying.

**Comment by essiw** [comment]
^I don't know how you can say that ARM JIT is possible for jailed devices, as @WilliamLCobb said in an other thread "I've been looking a ton into the iOS signing mechanisms and it looks like it's going to be near impossible to implement for jailed devices"... Please stop getting the hopes up for people by stating things like they are a fact when you don't know it for sure.

**Comment by DarknesGaming** [comment]
You are right. I don't know for sure if ARM JIT will be possible on Jailed devices. I probably shouldn't have use the word "will" in my last comment here.
@WilliamLCobb did say it's near impossible, so it's slightly possible for him to actually get it working. As long as he doesn't quit, it could happen.

**Comment by OkazakiTheOtaku** [comment]
Honestly, I think that it isn't as necessary. As time goes on, move devices
will be capable. The 6s series can run at full speed on any game. Heck, my
6s+ can run PPSSPP at near full speed without JIT. I don't think that
chasing stars is good at this point when there are issues that CAN be fixed
at the moment. At the very least, let's save JIT for when there is a
working jailbreak.

**Comment by DarknesGaming** [comment]
That's probably a good idea. There should be a jailbreak for iOS 9.3-9.3.2 in a few weeks, so we should save JIT for later.
Maybe we can find a way to add DeSmuME 0.9.11 as a seperate core instead of trying to replace the DeSmuME 0.9.10 core that already works with iOS.

**Comment by Razzile** [comment]
JIT was definitely possible pre iOS 9.2.1 because an exploit existed to disable codesigning which could let JIT run. I am looking into JIT as well but because of exams i don't have much time to do it

**Comment by DarknesGaming** [comment]
I'm going to start focusing on iNDS 2.0 now. I really want to be part of the iNDS team because I'm bored, and I am interested in coding.

**Comment by Razzile** [comment]
Same. After exams (end of june) i can probably help out much more

**Comment by DarknesGaming** [comment]
I am currently analyzing iNDS right now.
![2016-05-06](https://cloud.githubusercontent.com/assets/17773074/15090043/14776802-13dc-11e6-958d-4020d8436e0f.png)

**Comment by OkazakiTheOtaku** [comment]
I mean, it's an emulator. It's going to use a lot of resources. I don't know what you expected.

**Comment by DarknesGaming** [comment]
I was expecting WAY more power than that.

**Comment by DarknesGaming** [comment]
I am currently testing 7-zip 15.14 on iNDS. Currently, iNDS is using 7-zip 9.22 beta.

**Comment by WilliamLCobb** [comment]
I just ran a time profile on iNDS and saw only 32% of the execution time is actually spent emulating the CPU. The other 68% is emulating the graphics. So even if JIT was twice as fast as the interpreter, the speed difference wouldn't really be noticeable. I think focusing on implementing an actual OpenGL renderer would be more productive for performance and could lead to better graphics

**Comment by IcYGodZ** [comment]
If you play Mario And Luigi Partners in time, it lags bad. During the Level up scene, time warps, entering battles and leaving battles. Its a lagfest for a 64 bit device. I even try to lower the filter to low quality or none, it LAGS like no tomarrow. I even see alot of white flashes of lines "maybe they are called tearing". Everytime there is a wierd graphical issue, I can't see numbers texts because they are sticking with other texts or behind them. So many errors lol

**Comment by DarknesGaming** [comment]
Maybe DeSmuME for computer platforms work better than on mobile platforms. If that's the case, than the developers of DeSmuME seriously need to focus more on mobile platforms.

**Comment by DarknesGaming** [comment]
The mute button is fixed, and @WilliamLCobb updated iNDS to 1.6.0. He also fixed some other issues, and that means redownloading the source code and fixing those issues in iNDS 2.0.0. Luckily I got Notepad++ so this should be easy for me.

**Comment by DarknesGaming** [comment]
Has the frame skip option always been on iNDS or did I not notice it and thought that option wasn't there?
I just downloaded iNDS 1.4.0 on iEmulators and noticed that the option is there.

**Comment by IcYGodZ** [comment]
Maybe once 2.0.0 is released, we can focus on optimizations completely for a short while.

**Comment by DarknesGaming** [comment]
That's a great idea, except then we would have to make a small update called 2.0.1 for the optimizations, and just like the epic failure of iOS 8 (lots of bugs there), we don't want to have to make those who use BuildStore or iEmulators to wait a long time before that updated version comes out. Plus, that kind of ruins the whole "big release" thing.

**Comment by DarknesGaming** [comment]
I just did some research on the ARM architecture, both 32-bit and 64-bit and went through the ArmLJit.cpp file. It turns out that the 64-bit ARM architecture (ARMv8-A) is actually called AArch64, and is backwards compatible with the 32-bit ARM architecture (ARMv7-A). @WilliamLCobb just defined that as ARM64 in the cpp file. Maybe a little fix will work.

**Comment by Razzile** [comment]
@DarknesGaming no offence but I'm guessing everyone here knew this already, with it being an *iOS* emulator

**Comment by DarknesGaming** [comment]
Really? Everyone already knew that the 64-bit architecture of ARM is called AArch64? Not everyone knows that.
Since I made a bunch of comments, you have to be more specific in what everyone already knows.

**Comment by DarknesGaming** [comment]
I am really hoping that anything that was fixed in the master branch will be fixed in the 2.0 branch. I tried to send commits full of anything fixed in the master branch just so we don't lose focus on that.

**Comment by DarknesGaming** [comment]
We seem to be losing focus on iNDS 2.0.0. As much as I really like to be part of something, and that you all have lives and exams to get to (school is painful), we must not forget about the big release of iNDS.

**Comment by Razzile** [comment]
As much as I can see you want to help the project, there's no need to rush things. And ultimately it's up to @WilliamLCobb to choose what to implement and when.

**Comment by DarknesGaming** [comment]
You are right on that. I need to be more patient and not rush things.
I'm still testing new stuff on iNDS 2.0.0. This will take some time for me, and if everything works, I can send a pull request to @WilliamLCobb and maybe he'll accept it. I just need to do chunks at a time.

**Comment by DarknesGaming** [comment]
@WilliamLCobb was right. Upgrading parts of iNDS does require a lot of editing. And 7-zip 15.14 is full of errors.
I did manage to fix the TinyXML problem after I upgraded it to TinyXML2. It's not that hard.

**Comment by DarknesGaming** [comment]
iNDS 2.0.0 is starting to become more awesome. I should go ask @WilliamLCobb for permission to start rolling out dev builds for testing. And then start making pull requests that he can merge into his 2.0 branch just to keep development going.

**Comment by DarknesGaming** [comment]
My 2.0.0 branch took down a lot of bug fixes, forcing those bugs to return.
Time to delete the branch and make a new one. I really do have a problem.

**Comment by DarknesGaming** [comment]
Had to delete my 2.0.0 branch on my fork. I will make a new one once I grab the master source code and edit/upgrade/remove some stuff.
Development is hard. Doing hard things can be fun. Easy things are just too boring.

**Comment by K0m0** [comment]
I havent seen this or havent looked hard enough. Is it possible to add an In app updater thing? Kinda like what gba4ios has. I would like to update to the newest version without going to build.io everytime I read theres a new update and completely doing a fresh install.

**Comment by essiw** [comment]
@k0m0 you don't need to do a fresh install each time. Keep the app and just go to build.io and click on install, it will the  install the new version while keeping all the save data and games that you had installed.

**Comment by DarknesGaming** [comment]
Here is what I got so far:
The 2.0.0 branch on my fork has not been made yet. I will make it after @WilliamLCobb makes his.
I have successfully upgraded TinyXML to TinyXML2. This new version is made by the same developer, and is significantly faster.
A new DeSmuME engine is in the works. This will be version 0.9.11, and will be used as a secondary engine.
The LZMA SDK will be updated from 9.22 beta to 15.14. This code was patched for Mac OS X and iOS use by OlehKulykov, and is in Objective-C and Swift format. I'm still working on how to get this working on Xcode.
More stuff coming soon.

**Comment by DarknesGaming** [comment]
I managed to update the LZMA SDK to 15.14, but I had to add the InlineObjc source code to get it to function. Also, the reader mm file had to be excluded because I could NOT figure out what was wrong with the line with "prop { 0 };" and I got a message that said "Expected Expression". I don't know what that means!

**Comment by DarknesGaming** [comment]
iNDS 2.0.0 Dev Build 1 has been built. TinyXML2 is now functioning properly, I upgraded the LZMA SDK to 15.14, which replaces Mejong's edited code, and added OlehKulykov's patch code and Inlineobjc code to get this LZMA SDK to function.
The secondary DeSmuME engine has been added to the project, but not referenced by Xcode. I am configuring the engine so it can work on mobile devices. Once it's done, I will add that option in the Emulation Engine settings for testing. I strongly advise you do NOT use it for everyday use and gaming.
The LZMA Extractor code created by Mejong is giving off errors. Apparently, it has no idea what the arm64 architecture is, which makes sense because arm64 came out 2 years after Mejong's LZMA Extractor code.
That's all I got. And now I will return to my development.

**Comment by DarknesGaming** [comment]
LZMA Extractor is being added back, and GNU Lightning JIT will be in testing for iNDS 2.0.0 Dev. Both 32-bit and 64-bit devices will be testing version 2.1.0 of GNU Lightning JIT.
The DeSmuME 0.9.11 engine has not been configured yet, and I still don't know how to make iNDS switch core folders. I'm still waiting for @WilliamLCobb to remake the 2.0.0 branch because mine has zero updates.

**Comment by DarknesGaming** [comment]
I finally managed to build iNDS-Dev, but the app crashes when it tries to extract anything. I'll just have to revert back to LZMA SDK 9.22 beta from Mejong, or I can send iNDS-Dev to @WilliamLCobb and hope he can fix this.

**Comment by WilliamLCobb** [comment]
Ugrading LZMA is probably not worth the time since the one we have now work really well for what we use it for. Let me know if you get JIT working and I'll test and merge your fork. I'm probably going to update DeSmuME soon to 0.9.11 pretty soon. I changed the structure of DeSmuME 0.9.10 quite a bit so it would probably be best to let me handle that upgrade

DeSmuME 0.9.11 comes with a completely different JIT backend and an OpenGL renderer. Getting both of those to work could yield a 3 or 4 times speed increase

**Comment by mk8itra1n** [comment]
@WilliamLCobb Yeah I believe @DarknesGaming commented on another issue and he said he got JIT working for 64bit devices and he was waiting for a response from you to merge the commit

**Comment by DarknesGaming** [comment]
I did not send the commits yet, and I only added the option for 64-bit devices. Plus, summer time is the perfect time to start learning how to program and code stuff.
Also, I only point at myself when someone asks who is a noob at C++. That coding language has such a high learning curve.

**Comment by mk8itra1n** [comment]
@DarknesGaming That'll be fine. I got a 64-Bit device cuz I use iNDS on an iPhone 6 Plus

**Comment by DarknesGaming** [comment]
It has been almost a month since iNDS ever got an update from the time of this comment. Maybe I should learn the C programming language, but I have to learn Java from my uncle first. He's got a major in Java, and a few other programming languages.
Check out my Calculator [here.](https://github.com/DarknesGaming/Calculator) I've given up on trying to improve iNDS because of my lack of C programming skills.

**Comment by reigningchamp** [comment]
Hey @WilliamLCobb , no real connection to this project or coding of any sort, just someone who enjoys your work :). Just wanted to let ya know i appreciate everything you've put into this project and i look forward to desmume 0.9.11 integration.

**Comment by DarknesGaming** [comment]
I look forward to when iNDS 2.0.0 comes out. I might reappear later in time, since I don't know the C programming language.
The current iNDS project could use some new frameworks to replace the missing ones.
